### PR TITLE
Register and unregister fcm Tokens on sign in and sign out

### PR DESCRIPTION
**Pull request by TaiHaDev**
## Describe your changes
I implemented the feature to register and unregister FCM tokens upon login and sign out. 
Changes:

- Used the firebase messaging dependency and connect to our firebase fireapp project using flutter firebase CLI
- Created token class payload
- Some code to initialise the firebase class on start up in main.dart
- The rest all the logic code to handle register and unregister code on log in and sign out

### CODE REVIEW

**Comment by TaiHaDev** [comment]
Hi @BenMMcLean, it seems like the build process just failed because I added new dependencies: Firebase Messaging and Firebase Core. I did include them in the pubspec.yaml, and it works on my local machine, but it may not be working here. Could you give me some advice on what I should do about this? Thank you!

File: fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart
Comment (emilymclean) [inline fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart]: Why are you accepting userId as a parameter? Just import the user repository and get it from there

Comment (emilymclean) [reply fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart]: Also it should be token singular

File: fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart
Comment (emilymclean) [inline fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart]: This is a bad place to put this. Every time a user logs in you are resubscribing to token refresh. In a hypothetical case that a user signs in and out ten times in a continuous session, this would result in a token refresh sending the same token ten times with 10 different user ids. It would send them even if the user is signed out, and throw an exception that I'm pretty sure would result in an app crash, since the thread execution would have since moved on.

Comment (emilymclean) [reply fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart]: I think this would be better suited in the initState of MyApp (or even better in a ViewModel attached to MyApp)

File: fireapp/lib/main.dart
Comment (emilymclean) [inline fireapp/lib/main.dart]: Just put this in initState

File: fireapp/lib/main.dart
Comment (emilymclean) [inline fireapp/lib/main.dart:62]: Add a line after this one, functions should have at least a line between them

File: fireapp/lib/pages/settings/setting.dart
Comment (emilymclean) [inline fireapp/lib/pages/settings/setting.dart]: You've created a race condition between the authentication token being deleted and the request to unregister your FCM token. If logout occurs first, your request will fail. Just await the unregister call before calling logout and you should be good.

File: fireapp/lib/pages/settings/setting.dart
Comment (emilymclean) [inline fireapp/lib/pages/settings/setting.dart]: I understand this page is from the old world where the previous developers mixed business and view logic, but we shouldn't be adding to it. Please add a proper ViewModel. Don't bother migrating over anything but the logout button logic though, keep it minimal.

File: fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart
Comment (emilymclean) [reply fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart]: Since extracting this code means calling this function in the repository multiple different ways, I would extract out the case where you _don't_ already have a token into a use case (a use case is literally just a class that holds business logic. A viewmodel also does that, but the point of a use case is that multiple viewmodels can access the same logic without individually duplicating it + (more relevantly here and common in UML based design) it can help define the different ways data can be accessed) that overloads the [call operator](https://dart.dev/language/callable-objects). E.g
```dart
@injectable
class RegisterCurrentTokenUseCase {
  final NotificationTokenRepository _notificationTokenRepository;
  RegisterCurrentTokenUseCase(this._notificationTokenRepository);

  void call() async {
    // <get the token and pass it to the repository>
  }
}
```
You'd then update the repository to accept the token. LoginViewModel would take this UseCase instead of the repository, and then call it like `await _registerCurrentTokenUseCase()`. On the other hand, MyApp would interact with the repository directly, since it already has the token.

File: fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart
Comment (emilymclean) [inline fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart]: Nitpick, but this seems like something we may want to use elsewhere. Please create a DeviceRepository, make this a function within it, and then call that from here.

File: fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart
Comment (emilymclean) [reply fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart]: Put the UseCase in the domain folder, in a new subdirectory called use_cases

File: fireapp/lib/domain/models/fcm_tokens.dart
Comment (emilymclean) [inline fireapp/lib/domain/models/fcm_tokens.dart]: You only need to provide the token field, the rest are derived

File: fireapp/lib/data/client/api/rest_client.dart
Comment (emilymclean) [inline fireapp/lib/data/client/api/rest_client.dart]: Just a heads up, I've instructed Chim to change this endpoint to `POST /v2/user/<user-id>/token`, and the unregister one to `DELETE /v2/user/<user-id>/token`

File: fireapp/lib/domain/models/fcm_tokens.dart
Comment (emilymclean) [reply fireapp/lib/domain/models/fcm_tokens.dart]: Also this shouldn't be plural either

File: fireapp/lib/presentation/login/login_view_model.dart
Comment (emilymclean) [inline fireapp/lib/presentation/login/login_view_model.dart]: An await at the end of a function means nothing if you don't also await the parent function. Also something is funky with the whitespace, please figure out what code is indented incorrectly and fix it

**Review by emilymclean — CHANGES_REQUESTED** [review]
I didn't realise you were also working on this, so the changes I made to the pipelines to support Firebase Messages is unfortunately branched off Julius's work. Once his is merged, just merge in develop (and fix the conflicts) and it should work fine. In the meantime just address the feedback here.

File: fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart
Comment (TaiHaDev) [reply fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart]: Hi @BenMMcLean, thank you so much for the detailed feedback! I understand the bug you mentioned, but I'm not entirely sure about the use case, as MyApp doesn't interact with the repository. It makes sense to me to move the on-refresh token listener to the initState in MyApp, and let the LoginViewModel be responsible for posting the new FCM token to the database. Could you explain further how I should implement the use case here?

Comment (emilymclean) [reply fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart]: The code in MyApp _should be_ interacting with the repository. As a rule, code above the data layer (Domain classes like ViewModels and UseCases) should _never_ interact with client and persistence classes. It may not be obvious from our fairly simple usage of the pattern in the app, but the purpose of having a repository layer is to abstract away data fetching and/or storage. An example of this you can view would be the `AuthenticationRepository`. In it, it uses the client to login, and then transparently caches the details in the persistence class. This leaves the domain layer to be more purely focused on business logic, rather than the peculiarities of data retrieval and storage.

The reason I suggest a use case is because there are two different ways you are interacting with the repository - one where you already have a token (refresh), and one where you don't (login). The repository should expect a token to be provided to it, and then you pull the logic to retrieve a token out of the repository and put it into a use case. To be clear, **you should update the repository function to accept an FCM token**. You can think of a use case as a discrete piece of business logic. Then the LoginViewModel just has to call an injected instance of the use case, rather than managing it itself.

In terms of implementing it, just fill in the code I've written above (read the link to call operator if you don't understand invoking an object instance). It should just be a matter of getting the token, then calling the repository function with that token. Then you just need to inject it into `LoginViewModel` and invoke it where you were previously calling the repository function.

Comment (TaiHaDev) [reply fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart]: Hi @BenMMcLean,

Thank you for your detailed explanation, and I fully understand it. However, I am currently facing a problem where I tried to listen to token refresh from MyApp and update the database by calling a backend endpoint, but it's requiring both UserId and FCM token. The token is always available, but the UserId is not available at startup—it's only available after sign-in.

I can think of a solution to implement a simple fix in the backend endpoint or perform a null pointer check inside the lambda. Overall, I think the token refresh callback is not that important, as it is rarely triggered—only in extreme cases according to the documentation I found.

Comment (emilymclean) [reply fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart]: It's true it is only triggered in extreme cases, but we still need to implement it. Just check if `AuthenticationRepository.getCurrentSession()` is null or not in the lambda. Make sure the lambda is in a ViewModel though, more broadly it would be useful to have an application scoped viewmodel

Comment (TaiHaDev) [reply fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart]: The `AuthenticationRepository.getCurrentSession()` will only return a non-null value if the user is already logged in. This is because `getCurrentSession()` caches the session after the` logIn` function is triggered. Therefore, there are still cases where the user might not be logged in when the token refreshes.

The best approach I can think of is to use `getCurrentSession()` and only update the database through the backend if it returns a non-null value—meaning, only if the user is logged in. This logic makes sense to me.

What I plan to do next is create a new ViewModel class for main.dart to handle interactions with the FCM repository. I believe I'll need to delete the old token and add the new one during the token refresh process.

Please let me know if I'm on the right track! Thank you so much!

Comment (emilymclean) [reply fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart]: Yes that is broadly what I was suggesting. You do not need to delete the old token, the backend will know if it is invalid when it attempts to send a notification.

File: fireapp/lib/domain/models/fcm_tokens.dart
Comment (TaiHaDev) [reply fireapp/lib/domain/models/fcm_tokens.dart]: Hi @BenMMcLean, I have fixed the code based on your feedback, except for this part. By 'derived,' do you mean that I should write a constructor that extracts the state for all other fields except for the token (i.e., grab the userId from the authentication repository)?

Comment (emilymclean) [reply fireapp/lib/domain/models/fcm_tokens.dart]: No, I mean the server figures it out (except for device type, I forgot about that one). You're basing your model off the database model when you should be basing it off the [request model in the controller](https://github.com/TechlauncherFireApp/backend/blob/ded2c4f20c00f503de5860ad5be0fec811fec452/controllers/v2/fcm_tokens/api.py).

File: fireapp/lib/data/client/fcm_tokens_client.dart
Comment (emilymclean) [inline fireapp/lib/data/client/fcm_tokens_client.dart]: All these file names are still plural. Please rename

File: fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart
Comment (emilymclean) [inline fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart]: *Don't* accept FCMToken, take the token string and figure out the rest here

File: fireapp/lib/data/client/api/rest_client.dart
Comment (emilymclean) [inline fireapp/lib/data/client/api/rest_client.dart]: register and unregister do not take the same payload

File: fireapp/lib/main.dart
Comment (emilymclean) [inline fireapp/lib/main.dart]: Please revert these changes, this reformatting isn't relevant to your PR

File: fireapp/lib/pages/settings/setting.dart
Comment (emilymclean) [inline fireapp/lib/pages/settings/setting.dart]: This formatting isn't relevant to your PR

File: fireapp/lib/pages/settings/setting.dart
Comment (emilymclean) [inline fireapp/lib/pages/settings/setting.dart]: Everything in this function is business logic, put it in the view model

File: fireapp/lib/presentation/login/login_view_model.dart
Comment (emilymclean) [inline fireapp/lib/presentation/login/login_view_model.dart]: Undo the formatting

**Review by emilymclean — CHANGES_REQUESTED** [review]
Are a bunch of files missing from this change?

**Comment by TaiHaDev** [comment]
Hi @BenMMcLean,

I have implemented the changes you requested. I apologise for the earlier mistakes. I tend to have a bad habit of reformatting code after writing it.

I have tested both the sign-in and sign-out functionalities. The sign-in works well, but when I try to sign out, the backend receives the DELETE HTTP request but responds with an unauthorized (401) status. I’m currently investigating this issue, but it’s quite challenging. If you have any insights, please let me know!

Thank you so much!

File: fireapp/lib/pages/settings/setting.dart
Comment (TaiHaDev) [reply fireapp/lib/pages/settings/setting.dart]: The function was there before. Should I extract the entire function and put it in the ViewModel? I'm not quite sure about the navigation part.

Comment (emilymclean) [reply fireapp/lib/pages/settings/setting.dart]: Yes, extract all of it. The navigation can be handled the same way as in the login view model/view

**Comment by emilymclean** [comment]
@TaiHaDev I cannot provide advice because it appears that you have failed to push a significant chunk of your files, most notably in this case, the `SettingViewModel`. I do suspect your problems with DELETE are caused by a common mistake, but I would prefer not to speculate and mislead you. Please push *all* your work and I'll advise from there.

**Comment by TaiHaDev** [comment]
I am so sorry for the inconvenience. Somehow, it wasn't automatically added to version control. I have pushed all the necessary files. Thank you!

File: fireapp/lib/pages/settings/setting_view_model.dart
Comment (emilymclean) [inline fireapp/lib/pages/settings/setting_view_model.dart]: @TaiHaDev Yep it was what I suspected! Easy fix, easy mistake. You need to await this line. What is currently happening is that it starts running this function then immediately returns (i.e, doesn't wait for the call to finish). The next line in the function that calls this then deletes the auth token. By the time `unregisterToken` gets to making the call, the authentication details are no longer there. When you add an `await` before this, the function won't return until `unregisterToken` finishes.

Comment (TaiHaDev) [reply fireapp/lib/pages/settings/setting_view_model.dart]: Got it! I have fixed it below. Thanks for the detailed explanation.

File: fireapp/lib/domain/repository/notification_fcm_token_repository.dart
Comment (emilymclean) [inline fireapp/lib/domain/repository/notification_fcm_token_repository.dart]: I'm not really happy with this being here, and if it is it definitely shouldn't be public. Considering we often need to `getUserId`, I would suggest adding a helper function to AuthenticationRepository to get the userId (make the return nullable, like `getCurrentSession` is), then just inline the null check and exception.

Also please subclass Exception with a more specific one like `SignedOutException`. I don't really mind just using bare `Exception` in the ViewModel, since the only consumer is the UI. But in the repository, we should be returning more semantic exceptions since we may want to catch and handle them in some business logic layer.

File: fireapp/lib/domain/use_cases/register_token_use_case.dart
Comment (emilymclean) [inline fireapp/lib/domain/use_cases/register_token_use_case.dart]: Same note about the exceptions here.

File: fireapp/lib/domain/use_cases/register_token_use_case.dart
Comment (emilymclean) [inline fireapp/lib/domain/use_cases/register_token_use_case.dart]: Remove a line of whitespace here

File: fireapp/lib/main_view_model.dart
Comment (emilymclean) [inline fireapp/lib/main_view_model.dart:22]: Add a line of whitespace between lines 20 and 21, as well as line 21 and 22.

File: fireapp/lib/main_view_model.dart
Comment (emilymclean) [inline fireapp/lib/main_view_model.dart]: Yeah I think a .getUserId function is warranted in the authentication repository. I dunno why I didn't add one in the first place.

File: fireapp/lib/pages/settings/setting_view_model.dart
Comment (emilymclean) [inline fireapp/lib/pages/settings/setting_view_model.dart]: Don't do this. Do what the `LoginViewModel` does. If you're passing the `BuildContext` to the ViewModel its a sign you're doing something wrong.

**Review by emilymclean — CHANGES_REQUESTED** [review]
Ok, nearly there. Some minor feedback about exceptions, moving a function, and just change how you're doing navigation in SettingViewModel. Don't forget that you also need to merge in main at some point as well.

**Comment by TaiHaDev** [comment]
I created a new exception folder because I couldn't find an existing one. I am also trying to research how to implement navigation using the LoginViewModel.

File: fireapp/lib/domain/repository/authentication_repository.dart
Comment (emilymclean) [inline fireapp/lib/domain/repository/authentication_repository.dart]: Don't throw this here. Like I said, make userId nullable. Throw an exception in the FCM repository.

File: fireapp/lib/pages/settings/setting_navigation.dart
Comment (emilymclean) [inline fireapp/lib/pages/settings/setting_navigation.dart]: The name of this represents the destination, not the context in which it is called, so a better name would be `login()` and `LoginSettingNavigation`. Otherwise, you have implemented the navigation code entirely correctly!

File: fireapp/lib/pages/settings/setting_view_model.dart
Comment (emilymclean) [inline fireapp/lib/pages/settings/setting_view_model.dart]: I would name this `_signOutState` (and update the name of the other variable too). Theoretically this viewModel would control several independent operations, and it would therefore be better to have a state for each one.

**Comment by TaiHaDev** [comment]
Hi Ben, I have updated the class and variable names to follow the convention. it looks better now. Thank you!

**Review by emilymclean — DISMISSED** [review]
Looks good (aside from the whitespace I already commented on). Fix that, merge in main, and I'll re-approve and merge

**Comment by TaiHaDev** [comment]
Hi @BenMMcLean, I tried to merge and push, but it didn't pass the pipeline. I realized this happened because I didn't push the firebase_options.dart file. I didn't push it initially because it is generated by the FlutterFire CLI and contains the API key and related information to connect to our Firebase project.

I read some Git issues, and they mentioned that it's generally okay to push it since it's not considered vulnerable, so I pushed it to pass the pipeline. Please let me know if this is not okay. Thanks a lot!

**Comment by TaiHaDev** [comment]
❌ loading /home/runner/work/mobile/mobile/fireapp/test/presentation/login/login_viewmodel_test.dart (failed)
There's an error like this within the unit test pipeline. I don't think there's a file like that yet so it failed.

File: fireapp/lib/main.dart
Comment (emilymclean) [inline fireapp/lib/main.dart:36]: I'm assuming these two lines should still be here? Please add back. Keep the dotenv stuff out and remove its dependency though.

**Review by emilymclean — DISMISSED** [review]
Ok this is quite good. Add back those two lines, add firebase_options.dart to the .gitignore, and fix that unit test and I will merge

File: .gitignore
Comment (emilymclean) [inline .gitignore]: This does not ignore the file

File: fireapp/pubspec.lock
Comment (emilymclean) [inline fireapp/pubspec.lock]: The fact all these dependencies were removed and the flutter sdk lowered from 3.16.0 to 3.7.0 should tell you something went wrong with the lock. Please ensure you are using `fvm` when running flutter commands, do not just click on the IDE option (it will use whatever flutter you have on your system).

File: fireapp/lib/main.dart
Comment (emilymclean) [inline fireapp/lib/main.dart:37]: You still need to add back this line...

Comment (emilymclean) [reply fireapp/lib/main.dart:37]: After `.initializeApp` is called, to be clear

**Review by emilymclean — CHANGES_REQUESTED** [review]
Please correct the `.gitignore` and fix the `pubspec.lock` file.

**Comment by TaiHaDev** [comment]
I really appreciate all of your feedback on this pull request. I learned a lot from it, especially about git versioning, and I realised how careless I have been with these changes due to my lack of experience. Thank you so much for your patience and help 🙏

**Review by emilymclean — DISMISSED** [review]
I have approved after pushing a fix for the remaining unaddressed feedback, please review my changes in your own time.

**Review by emilymclean — DISMISSED** [review]

**Review by emilymclean — DISMISSED** [review]

**Review by emilymclean — APPROVED** [review]

**Pull request merged**
